src/Sarif/Readers/DeferredDictionary.cs

```diff
@@ -0,0 +1,514 @@
+    {
+        private JsonSerializer _jsonSerializer;
+        private Func<Stream> _streamProvider;
+        private long _start;
+
+        private Stream _stream;
+        private Dictionary<int, long> _itemPositions;
+
+        public DeferredDictionary(JsonSerializer jsonSerializer, JsonPositionedTextReader reader, bool buildPositionsNow = false)
+        {
+            _jsonSerializer = jsonSerializer;
+            _streamProvider = reader.StreamProvider;
+            _start = reader.TokenPosition;
+
+            // We have the JsonTextReader, which must scan to after the collection to resume building the outer object
+            // We may as well make the map of element positions.
+            if (buildPositionsNow)
+            {
+                BuildPositions(reader, 0);
+            }
+            else
+            {
+                reader.Skip();
+            }
+        }
+
+        private void EnsurePositionsBuilt()
+        {
+            if (_itemPositions == null) BuildPositions();
+        }
+
+        private void BuildPositions()
+        {
+            using (Stream stream = _streamProvider())
+            using (JsonPositionedTextReader reader = new JsonPositionedTextReader(() => stream))
+            {
+                stream.Seek(_start, SeekOrigin.Begin);
+                reader.Read();
+
+                BuildPositions(reader, _start);
+            }
+        }
+
+        private void BuildPositions(JsonPositionedTextReader reader, long currentOffset)
+        {
+            Dictionary<int, long> result = new Dictionary<int, long>();
+            while (true)
+            {
+                // Find the position just before the PropertyName (we need to read it back later to confirm the string matches)
+                long keyPosition = currentOffset + reader.TokenPosition + 1;
+
+                reader.Read();
+                if (reader.TokenType == JsonToken.EndObject) break;
+
+                if (reader.TokenType != JsonToken.PropertyName) throw new InvalidDataException($"@({reader.LineNumber}, {reader.LinePosition}): Expected property name, found {reader.TokenType} \"{reader.Value}\".");
+
+                // Read JSON object name (Dictionary key)
+                string key = (string)reader.Value;
+                int keyHash = key.GetHashCode();
+
+                // Skip the value
+                reader.Read();
+                reader.Skip();
+
+                // Add the hash to the position to our Dictionary; resolve collisions by incrementing
+                while (result.ContainsKey(keyHash))
+                {
+                    keyHash++;
+                }
+                result[keyHash] = keyPosition;
+            }
+
+            _itemPositions = result;
+        }
+
+        public T this[string key]
+        {
+            get
+            {
+                T value;
+                if (!this.TryGetValue(key, out value)) throw new KeyNotFoundException(key);
```

Comment:
>if (!this.TryGetValue(key, out value)) throw new KeyNotFoundException(key); [](start = 16, length = 75)

you can skip the var declaration and put this in a single line now by declaring the var inside the TryGetValue call, e.g.,

if (!this.TryGetValue(key, out T value)) { throw new KeyNotFoundException(key); }

Reply:
Huh - cool. I'll start using that pattern.